File: python/socketd/socketd/transport/core/stream/RequestStream.py
from abc import ABC
from typing import Coroutine, Callable

from socketd.transport.core.Costants import Constants
from socketd.transport.core.Entity import Reply
from socketd.transport.core.stream.StreamBase import StreamBase
from socketd.transport.utils.CompletableFuture import CompletableFuture


class RequestStream(StreamBase):

    def __init__(self, sid: str, timeout: int):
        super().__init__(sid, Constants.DEMANDS_SINGLE, timeout)
        self._future: CompletableFuture = CompletableFuture()

    def is_done(self):
        return self._future.done()

    def __await__(self):
        return self._future.get(self.timeout())

    def await_result(self):
        return self.__await__()

    async def on_reply(self, message: Reply):
        return self._future.set_result(message)

    def then_reply(self, onReply: Callable[[Reply], None]):
        self._future.then_callback(onReply)

File: go/socketd/transport/stream/impl/request_stream.go
package impl

import (
	"fmt"
	"time"

	"socketd/transport/stream"

	"socketd/transport/core/constant"
	"socketd/transport/core/message"
)

type RequestStream struct {
	*StreamBase //继承

	doOnReply func(message *message.Message)
	replyChan chan *message.Message
	done      bool
}

func NewRequestStream(sid string, timeout time.Duration) *RequestStream {
	return &RequestStream{StreamBase: NewStreamBase(sid, constant.DEMANDS_SIGNLE, timeout), replyChan: make(chan *message.Message)}
}

func (s *RequestStream) IsDone() bool {
	return s.done
}

func (s *RequestStream) OnReply(frame *message.Frame) {
	s.done = true
	s.replyChan <- frame.Message
}

func (s *RequestStream) Await() (*message.Message, error) {
	timer := time.NewTimer(time.Millisecond * 30) // 30毫秒超时
	for {
		select {
		case <-timer.C:
			return nil, fmt.Errorf("request reply timeout, sid = %s", s.StreamBase.Sid())
		case reply := <-s.replyChan:
			return reply, nil
		}
	}
}

func (s *RequestStream) ThenReply(onReply func(message *message.Message)) stream.RequestStream {
	s.doOnReply = onReply
	return s
}
